Handle comparing ChronoLocalDateTime and Instants

robdaemon opened the pull request:
During fingerprinting, we can end up trying to compare two datatypes
that are not comparable - ChronoLocalDateTime and Instant. Coerce the
ChronoLocalDateTime to a proper Instant for the comparison to work.

Resolves #12311

[ci all]

Commit:
Handle comparing ChronoLocalDateTime and Instants

During fingerprinting, we can end up trying to compare two datatypes
that are not comparable - ChronoLocalDateTime and Instant. Coerce the
ChronoLocalDateTime to a proper Instant for the comparison to work.

When earliest / latest are called, the acc parameter must be coerced to
a Temporal object also.

Resolves #12311

[ci all]

robdaemon commented on src/metabase/sync/analyze/fingerprint/fingerprinters.clj:
```diff
@@ -185,6 +186,14 @@
+  ;; to work with. Ideally, we would use the database's ZoneOffset, but
+  ;; we don't have access to that here. Use the JVM's systemDefault to
+  ;; convert this.
+  ChronoLocalDateTime (->temporal [this] (.toInstant this
```
if someone has an idea of the best way to get the database's timezone here, I'm all ears

sbelak replied:
Have a look at `query_processor.timezone` if you can use something from there

robdaemon replied:
I started with that, but it complains about the query processor not being initialized. I'm not sure how to do that.

sbelak replied:
I think that's only with tests, when this runs for real it's a part of the QP and the store should be initialized already (just reverse engineer which bits of the QP store you need for it to work).

sbelak reviewed (changes requested):
I'm not sure this fully fixes the problem. Dates as strings and dates as numbers still go through different constructors which return incompatible types, so the implementations for `Long` and `Integer` either need to recur so they go through `ChronoLocalDateTime`, or we repeat the same logic there.

sbelak commented on test/metabase/sync/analyze/fingerprint/fingerprinters_test.clj:
```diff
@@ -11,14 +11,32 @@
+                                     :latest   "2018-01-01T04:04:00Z"}}}
+           (transduce identity
+                      (fingerprinter (field/map->FieldInstance {:base_type :type/Temporal}))
+                      [(java.time.LocalDateTime/of 2013 01 01 20 04 0 0)
```
It would be good to test the code paths from numbers and strings, not just directly inserting the problematic types.

robdaemon commented:
@sbelak I think I addressed your concern about mixing numbers and strings, but I will need some help with getting the qp.store setup to get the database's timezone.

sbelak commented on src/metabase/sync/analyze/fingerprint/fingerprinters.clj:
```diff
@@ -157,12 +158,14 @@
    (some-> acc u.date/format))
   ([acc t]
-   (if (and t acc (t/before? t acc))
+   (if (and t acc (t/before? t (->temporal acc)))
```
I don't think this is a good solution -- if I change the order from string, number to number, string it breaks. I think it would be better to change implementations of `ITemporalCoerceable ` so that they always return the same thing. Something like: `Long     (->temporal [this] (->temporal (t/instant this)))`

Commit:
making ->temporal recursive

Commit:
Get the timezone in an precedence order

When converting a LocalTime for fingerprinting, convert to an Instant
using the following precedence order:

* report timezone
* database timezone
* system timezone

[ci all]

Commit:
use UTC for timezones

camsaul reviewed (approved):
cool  works for me. Don't forget to set the *Milestone* for the issue this closes.

sbelak commented on src/metabase/sync/analyze/fingerprint/fingerprinters.clj:
```diff
@@ -182,9 +186,21 @@
-  String   (->temporal [this] (u.date/parse this))
-  Long     (->temporal [this] (t/instant this))
-  Integer  (->temporal [this] (t/instant this))
+  String   (->temporal [this] (->temporal (u.date/parse this)))
```
Does `String` need recasting as well?

sbelak commented on src/metabase/sync/analyze/fingerprint/fingerprinters.clj:
```diff
@@ -182,9 +186,21 @@
+                                                         (getRules)
+                                                         (getOffset (java.time.Instant/now)))))
+  ChronoZonedDateTime (->temporal [this] (.toInstant this))
   Temporal (->temporal [this] this))
```
Is this compatible with all the others?

robdaemon replied:
it seems like it?

sbelak reviewed (approved):
LGTM, I'd sleep better if we had a test that tests if all 6 possible types mash together though.

robdaemon commented:
@sbelak okay I can add that test
